// @openfga/sdk/dist/configuration.js
"use strict";
/**
 * JavaScript and Node.js SDK for OpenFGA
 *
 * API version: 0.1
 * Website: https://openfga.dev
 * Documentation: https://openfga.dev/docs
 * Support: https://discord.gg/8naAwJfWN6
 * License: [Apache-2.0](https://github.com/openfga/js-sdk/blob/main/LICENSE)
 *
 * NOTE: This file was auto generated by OpenAPI Generator (https://openapi-generator.tech). DO NOT EDIT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = exports.GetDefaultRetryParams = void 0;
const axios_1 = require("axios");
const credentials_1 = require("./credentials");
const errors_1 = require("./errors");
const validation_1 = require("./validation");
function GetDefaultRetryParams(maxRetry = 3, minWaitInMs = 100) {
    return {
        maxRetry: maxRetry,
        minWaitInMs: minWaitInMs,
    };
}
exports.GetDefaultRetryParams = GetDefaultRetryParams;
class Configuration {
    constructor(params = {}, axios = axios_1.default) {
        this.axios = axios;
        /**
         * provide scheme (e.g. `https`)
         *
         * @type {string}
         * @memberof Configuration
         */
        this.apiScheme = "https";
        /**
         * Returns the API base path (apiScheme+apiHost)
         */
        this.getBasePath = () => `${this.apiScheme}://${this.apiHost}`;
        this.apiScheme = params.apiScheme || this.apiScheme;
        this.apiHost = params.apiHost;
        this.storeId = params.storeId;
        const credentialParams = params.credentials;
        if (credentialParams) {
            switch (credentialParams === null || credentialParams === void 0 ? void 0 : credentialParams.method) {
                case credentials_1.CredentialsMethod.ApiToken:
                    this.credentials = {
                        method: credentialParams.method,
                        config: {
                            token: credentialParams.config.token,
                            headerName: "Authorization",
                            headerValuePrefix: "Bearer",
                        }
                    };
                    break;
                case credentials_1.CredentialsMethod.ClientCredentials:
                    this.credentials = {
                        method: credentials_1.CredentialsMethod.ClientCredentials,
                        config: {
                            // We are only copying them from the passed in params here. We will be validating that they are valid in the Credentials constructor
                            clientId: credentialParams.config.clientId,
                            clientSecret: credentialParams.config.clientSecret,
                            apiAudience: credentialParams.config.apiAudience,
                            apiTokenIssuer: credentialParams.config.apiTokenIssuer,
                        }
                    };
                    break;
                case credentials_1.CredentialsMethod.None:
                default:
                    this.credentials = { method: credentials_1.CredentialsMethod.None };
                    break;
            }
        }
        const baseOptions = params.baseOptions || {};
        baseOptions.headers = baseOptions.headers || {};
        if (typeof process === "object" && process.title === "node" && !baseOptions.headers["User-Agent"]) {
            baseOptions.headers["User-Agent"] = "openfga-sdk {sdkId}/{packageVersion}".replace("{sdkId}", "js").replace("{packageVersion}", Configuration.sdkVersion);
        }
        this.baseOptions = baseOptions;
        this.retryParams = params.retryParams;
    }
    /**
     *
     * @return {boolean}
     * @throws {FgaValidationError}
     */
    isValid() {
        var _a;
        (0, validation_1.assertParamExists)("Configuration", "apiScheme", this.apiScheme);
        (0, validation_1.assertParamExists)("Configuration", "apiHost", this.apiHost);
        if (!(0, validation_1.isWellFormedUriString)(this.getBasePath())) {
            throw new errors_1.FgaValidationError(`Configuration.apiScheme (${this.apiScheme}) and Configuration.apiHost (${this.apiHost}) do not form a valid URI (${this.getBasePath()})`);
        }
        if (((_a = this.retryParams) === null || _a === void 0 ? void 0 : _a.maxRetry) && this.retryParams.maxRetry > 5) {
            throw new errors_1.FgaValidationError("Configuration.retryParams.maxRetry exceeds maximum allowed limit of 5");
        }
        return true;
    }
    /**
     * Check if the given MIME is a JSON MIME.
     * JSON MIME examples:
     *   application/json
     *   application/json; charset=UTF8
     *   APPLICATION/JSON
     *   application/vnd.company+json
     * @param mime - MIME (Multipurpose Internet Mail Extensions)
     * @return True if the given MIME is JSON, false otherwise.
     */
    isJsonMime(mime) {
        // eslint-disable-next-line no-control-regex
        const jsonMime = new RegExp("^(application/json|[^;/ \t]+/[^;/ \t]+[+]json)[ \t]*(;.*)?$", "i");
        return mime !== null && (jsonMime.test(mime) || mime.toLowerCase() === "application/json-patch+json");
    }
}
exports.Configuration = Configuration;
/**
 * Defines the version of the SDK
 *
 * @private
 * @type {string}
 * @memberof Configuration
 */
Configuration.sdkVersion = "0.0.2";
