// @auth0/fga/dist/base.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * JavaScript and Node.js SDK for Auth0 Fine Grained Authorization (FGA)
 *
 * API version: 0.1
 * Website: https://fga.dev
 * Documentation: https://docs.fga.dev
 * Support: https://discord.gg/8naAwJfWN6
 * License: [MIT](https://github.com/auth0-lab/fga-js-sdk/blob/main/LICENSE)
 *
 * NOTE: This file was auto generated by OpenAPI Generator (https://openapi-generator.tech). DO NOT EDIT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseAPI = exports.COLLECTION_FORMATS = void 0;
// Some imports not used depending on template conditions
// @ts-ignore
const axios_1 = require("axios");
const configuration_1 = require("./configuration");
/**
 *
 * @export
 */
exports.COLLECTION_FORMATS = {
    csv: ",",
    ssv: " ",
    tsv: "\t",
    pipes: "|",
};
/**
 *
 * @export
 * @class BaseAPI
 */
class BaseAPI {
    constructor(configuration, axios = axios_1.default) {
        this.axios = axios;
        if (configuration instanceof configuration_1.Configuration) {
            this.configuration = configuration;
        }
        else {
            this.configuration = new configuration_1.Configuration(configuration, axios);
        }
        this.configuration.isValid();
    }
}
exports.BaseAPI = BaseAPI;
